README.md
STAC/Geoserver Platform
=======================

Services
--------

    * ``stac_manager`` (port 8888), primary STAC service with functions to ``publish``, ``fetch``, ``delete``, STAC assets. Includes ``/list`` search console and ``/view`` layer previewer.

    * ``geoserver`` (port 8080), serves WMS tiles and used by STAC manager to register new assets

    * ``nginx`` (port 80, 443), two configurations:
        * stac_api_nginx_conf.d
            * reverse proxy to map ``https://../api`` to ``stac_manager:8888/api`` and ``https://../geoserver`` to ``geoserver:8080/geoserver``
        * geoserver_node_conf.d
            * maps ``http://../geoserver`` to ``geoserver:8080/geoserver`` service


    * ``db`` (port 5432), PostgreSQL database. Used by ``stac_manager`` for authentication and STAC catalog storage.
        * username and password given by ``DB_PASSWORD``, ``DB_USER``.
        * host volume mount at ``POSTGRES_VOLUME``

    * ``expand_drive_service`` (port ``DISK_RESIZE_SERVICE_PORT``), used to inrease the size of the disk used by the Geoserver node.
         * ``/resize``, ``POST``, ``{'gb_to_add': '12'}``

Configuration
-------------

When deploying the platform you must create a file call ``stac_envs``. This file sets environment variables for the Docker containers that allow them to operate correctly. An example file is given in the root of this project at ``example_stac_envs``.

    * ``DISK_RESIZE_SERVICE_ZONE`` -- GCE service zone for the disk
    * ``DISK_RESIZE_SERVICE_DISK_NAME`` -- the name of the disk in the GCE console/context.
    * ``DISK_RESIZE_SERVICE_MAX_SIZE_GB`` -- max allowed size of the disk in GB
    * ``DISK_RESIZE_SERVICE_DEVICE_NAME`` -- the name of the device as is it mounted locally to the host (i.e. ``/dev/sdb``)
    * ``DISK_RESIZE_SERVICE_ACCOUNT_KEYFILE`` -- a path to a GCE ``.json`` keyfile that has the following permissions enabled for the given disk:
        * ``compute.disks.create`` on the project to be able to create a new disk
        * ``compute.instances.attachDisk`` on the VM instance
        * ``compute.disks.use`` permission on the disk to attach


HISTORY.rst
Release History
===============

0.5.0 (2021/03/29)
------------------
* Changed functionality of "``--reduce_factor``" to operate on wildcard file
  list and modified the last argument to be a file suffix rather than a
  target file name.

0.4.1 (2020/12/07)
------------------
* Fixed issue with "``hash_val`` is not defined".

0.4.0 (2019/04/13)
------------------
* Changed behavior of command line function, now takes one of two primary
  commands ``process`` or ``publish``. The command ``process`` behaves like
  the original command line utility with same arguments and behavior. The
  ``publish`` command takes a Google Bucket URI, host to an ecoshard server,
  and an api key and published the raster to that server.
* Fixed an issue when summing up that would ignore nodata values.
* Fixes an issue where the overview interpolation argument was ignored.
* Added a ``download_and_unzip`` function to the API.

0.3.3 (2019/11/09)
------------------
* Fixed an issue that would cause the download rate to be under estimated.

0.3.1 (2019/10/07)
------------------
* Fixing an issue in download_to_url where the file might not be flushed and the
  final log message is not printed.

0.3.0 (2019/09/26)
------------------
* Added a new command line mode –reduce_factor that reduces the number of pixels
  in a raster by that integer amount. Ex:

    ``python -m ecoshard base.tif --reduce_factor 4 max target.tif``

    this call makes the size of the pixels in base.tif 4 times larger on the
    edge, thus reducing the total size of the image by 16 times, the convolution
    upsample is done with a "max" and the output file is ``target.tif``. The
    modes "min", "max", "sum", "average", "mode" are available.

0.2.2 (2019/09/24)
------------------
* Added a ``download_url`` function to ``ecoshard`` to fetch files via
  http(s).
* Developing Flask module to visualize ecoshards.

0.2.1 (2019/06/28)
------------------
* Hotfix: gs to gsutil for copying to buckets.

0.2.0 (2019/06/28)
------------------
* Added functionality to download a url.
* Added functionality to copy to a google bucket.
* Fixed an issue on functions that should write a token file but didn't.

0.1.1 (2019/06/27)
------------------
* Added a ``--force`` flag to the command line utility.

0.0.2 (2019/06/26)
------------------
* Initial release.


geoserver_service.py
"""Service to monitor for new snapshots and create disk & flush if so."""
import argparse
import datetime
import logging
import os
import socket
import subprocess
import sys
import time
import threading
import urllib
import uuid

import flask
import retrying

APP = flask.Flask(__name__)

logging.basicConfig(
    level=logging.DEBUG,
    format=(
        '%(asctime)s (%(relativeCreated)d) %(processName)s %(levelname)s '
        '%(name)s [%(funcName)s:%(lineno)d] %(message)s'),
    stream=sys.stdout)
LOGGER = logging.getLogger(__name__)
logger = logging.getLogger('waitress')
logger.setLevel(logging.DEBUG)

LAST_SNAPSHOT_NAME = None
LAST_DISK_NAME = None
HEALTHY = False


@retrying.retry(
    wait_exponential_multiplier=1000, wait_exponential_max=5000,
    stop_max_attempt_number=5)
def do_rest_action(
        session_fn, host, suburl, data=None, json=None, headers=None):
    """A wrapper around HTML functions to make for easy retry.

    Args:
        sesson_fn (function): takes a url, optional data parameter, and
            optional json parameter.

    Returns:
        result of `session_fn` on arguments.

    """
    try:
        return session_fn(
            urllib.parse.urljoin(host, suburl), data=data, json=json,
            headers=headers)
    except Exception:
        LOGGER.exception('error in function')
        raise


def new_disk_monitor_docker_manager(
        snapshot_pattern, mount_point, check_time, image_name, container_name,
        mem_size):
    """Monitor for new snapshots matching pattern. Create copy if one appears.

    Args:
        snapshot_pattern (str): the string to match for a new snapshot
        mount_point (str): system location to mount the drive
        check_time (float): how many seconds to wait between snapshot checks
        image_name (str): name of image to run for Docker container
        container_name (str): what to name the running docker container
        mem_size (str): Java formatted max memory string i.e. "12G"

    Returns:
        None

    """
    global HEALTHY
    global STATUS_STRING
    global LAST_SNAPSHOT_NAME
    global LAST_DISK_NAME
    disk_iteration = 0
    container_running = False
    while True:
        try:
            # get existing devices
            lsblk_result = subprocess.run(
                ["lsblk"], stdout=subprocess.PIPE,
                check=True).stdout.rstrip().decode('utf-8')
            existing_dev_names = set([
                line.split(' ')[0] for line in lsblk_result.split('\n')])

            STATUS_STRING = 'search for new snapshot'
            LOGGER.info(STATUS_STRING)
            snapshot_query = subprocess.run([
                "gcloud", "compute", "snapshots", "list", "--limit=1",
                "--sort-by=~name", f"--filter=name:({snapshot_pattern})",
                "--format=value(name)"], stdout=subprocess.PIPE,
                check=True)
            snapshot_name = snapshot_query.stdout.rstrip().decode('utf-8')
            LOGGER.debug(f"this is the latest snapshot: {snapshot_name}")
            if (snapshot_name == LAST_SNAPSHOT_NAME and
                    LAST_SNAPSHOT_NAME is not None):
                # not a new snapshot
                STATUS_STRING = f'last checked: {str(datetime.datetime.now())}'
                LOGGER.info(STATUS_STRING)
                continue
            STATUS_STRING = f'creating new disk from snapshot {snapshot_name}'
            LOGGER.info(STATUS_STRING)

            # create new disk
            hostname = socket.gethostname()
            disk_name = (f'{hostname}-data-{uuid.uuid4().hex}')[:59]
            STATUS_STRING = f'creating disk {disk_name}'
            LOGGER.info(STATUS_STRING)
            disk_iteration += 1
            subprocess.run([
                "gcloud", "compute", "disks", "create", disk_name,
                f"--source-snapshot={snapshot_name}", "--type=pd-ssd",
                "--zone=us-west1-b"], check=True)

            # attach the new disk to the current host, sometimes it takes a bit
            # for the disk to become available to attach after it's been
            # created, be tolerant of that
            attach_attempts = 0
            STATUS_STRING = f'attaching disk {disk_name}, inital attempt'
            LOGGER.info(STATUS_STRING)
            while True:
                # give it 5 seconds to come online
                time.sleep(5)
                try:
                    attach_attempts += 1
                    subprocess.run([
                        "gcloud", "compute", "instances", "attach-disk",
                        hostname, f"--disk={disk_name}", "--zone=us-west1-b"],
                        check=True)
                    STATUS_STRING = \
                        f'attached {disk_name}, {attach_attempts} attempts'
                    LOGGER.info(STATUS_STRING)
                    break
                except subprocess.CalledProcessError:
                    STATUS_STRING = f'attach attempt {attach_attempts} failed'
                    LOGGER.exception(STATUS_STRING)
                    if attach_attempts > 10:
                        HEALTHY = False
                        raise RuntimeError(
                            f'{attach_attempts} for disk {disk_name} failed')

            STATUS_STRING = f'setting disk {disk_name} to autodelete'
            LOGGER.info(STATUS_STRING)
            subprocess.run([
                "gcloud", "compute", "instances", "set-disk-auto-delete",
                hostname, f"--disk={disk_name}", "--zone=us-west1-b"],
                check=True)

            # unmount the current disk if any is mounted
            try:
                STATUS_STRING = f'unmounting {mount_point}'
                LOGGER.info(STATUS_STRING)
                subprocess.run(["umount", mount_point], check=True)
            except Exception:
                LOGGER.exception(f'exception when unmounting {mount_point}')

            STATUS_STRING = f'ensuring {mount_point} exists'
            LOGGER.info(STATUS_STRING)
            subprocess.run(["mkdir", "-p", mount_point])

            LAST_SNAPSHOT_NAME = snapshot_name

            # mount the new disk at the mount point
            lsblk_result = subprocess.run(
                ["lsblk"], stdout=subprocess.PIPE,
                check=True).stdout.rstrip().decode('utf-8')
            new_dev_names = set([
                lsblk_result.split(' ')[0]
                for lsblk_result in lsblk_result.split('\n')])
            mount_device = list(new_dev_names.difference(
                existing_dev_names))[0]
            device_location = f'/dev/{mount_device}'
            STATUS_STRING = f'mounting {device_location} at {MOUNT_POINT}'
            LOGGER.info(STATUS_STRING)
            try:
                subprocess.run(
                    ["mount", device_location, MOUNT_POINT], check=True)
            except Exception:
                LOGGER.exception("mount failed")
                HEALTHY = False
                raise

            try:
                if container_running:
                    STATUS_STRING = f'stopping docker container'
                    LOGGER.info(STATUS_STRING)
                    subprocess.run(
                        ["docker", "stop", container_name], check=True)
                    container_running = False

                STATUS_STRING = f'starting docker container'
                LOGGER.info(STATUS_STRING)
                subprocess.run([
                    "docker", "run", "--rm", "-d", "-it", "-v",
                    "/mnt/geoserver_data/:/usr/local/geoserver/data_dir",
                    "-p", "8080:8080", "--name", container_name,
                    image_name, "api.salo.ai", "8888", "maps.salo.ai", "8080",
                    mem_size], check=True)
                container_running = True
                HEALTHY = True
            except Exception:
                LOGGER.exception('stopping or starting failed')
                HEALTHY = False

            # Detach and delete the old disk
            if LAST_DISK_NAME:
                STATUS_STRING = f'detatch old {LAST_DISK_NAME}'
                LOGGER.info(STATUS_STRING)
                subprocess.run([
                    "gcloud", "compute", "instances", "detach-disk",
                    hostname, f"--disk={LAST_DISK_NAME}", "--zone=us-west1-b"],
                    check=True)

                STATUS_STRING = f'deleting old {LAST_DISK_NAME}'
                LOGGER.info(STATUS_STRING)
                subprocess.run([
                    f"yes|gcloud compute disks delete {LAST_DISK_NAME} "
                    "--zone=us-west1-b"], check=True, shell=True)

            LAST_DISK_NAME = disk_name
            STATUS_STRING = f'last checked: {str(datetime.datetime.now())}'

        except Exception as e:
            STATUS_STRING = f'error: {str(e)}'
            LOGGER.exception(STATUS_STRING)
        LOGGER.debug(f'sleeping {check_time} seconds')
        time.sleep(check_time)


@APP.route('/', methods=['GET'])
def processing_status():
    """Return the state of processing."""
    if HEALTHY:
        return (
            f'last snapshot: {LAST_SNAPSHOT_NAME}\n'
            f'last disk: {LAST_DISK_NAME}\nstatus: {STATUS_STRING}', 200)
    return STATUS_STRING, 500


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='disk remounter')
    parser.add_argument(
        '--app_port', type=int, required=True,
        help='port to respond to status queries')
    parser.add_argument(
        '--snapshot_pattern', type=str, required=True, help=(
            'pattern of disk snapshot to query for, ex '
            '"geoserver-data-disk*"'))
    parser.add_argument(
        '--mount_point', type=str, required=True, help=(
            'mount point location ex /mnt/geoserver_data'))
    parser.add_argument(
        '--check_time', type=float, default=5*60, help=(
            'how many seconds to wait between checking for a new disk'))
    parser.add_argument(
        '--container_name', type=str, required=True, help=(
            'desired name of docker container running the geoserver node'))
    parser.add_argument(
        '--image_name', type=str, required=True, help=(
            'docker image name to run'))
    parser.add_argument(
        '--mem_size', type=str, required=True, help=(
            'java formatted max mem size string eg "12G"'))

    parser.add_argument('--initalize', action="store_true")
    args = parser.parse_args()
    snapshot_PATTERN = args.snapshot_pattern
    MOUNT_POINT = args.mount_point
    DISK_ITERATION = 0
    STATUS_STRING = "startup"
    PASSWORD_FILE_PATH = os.path.join(
        args.mount_point, 'data', 'secrets', 'adminpass')
    swap_thread = threading.Thread(
        target=new_disk_monitor_docker_manager,
        args=(args.snapshot_pattern, args.mount_point, args.check_time,
              args.image_name, args.container_name, args.mem_size))
    swap_thread.start()

    APP.run(
        host='0.0.0.0',
        port=args.app_port)
